### DISCUSSION TITLE
spawn clients in specific window in specific tag after starup

### DISCUSSION THREAD

**Discussion by RayZ0rr** [herbstluftwm/herbstluftwm#1425] [Q&A]
The setup I want (commands entered are shown in center) : 
In tag 1
```
+-Win 1--------+----Win 2----+
|              |            |
|  neofetch    |    la      |
|              |            |
|              |            |
+--------------+------------+
```
In tag 2
```
+------Title Vifm---------+
|                        |
|          vifm          |
|                        |
|                        |
+------------------------+
```
In tag 3
```
+----Title Bottom-------+
|                        |
|          btm           |
|                        |
|                        |
+------------------------+
```

My startup script : 
```bash
#!/usr/bin/env bash

# Requirement: bash >= 4.0 (because of the usage of $BASHPID)

spawn_with_rules() {(
    # this rule also requires, that the client
    # sets the _NET_WM_PID property
    # herbstclient rule once pid=$BASHPID maxage=10 "${RULES[@]}"
    herbstclient rule once pid=$BASHPID "${RULES[@]}"
    exec "$@"
    ) &
}

# spawn two terminals (alacritty and default terminal in two windows) in tag 1
RULES=( tag=1 )
spawn_with_rules alacrity
sleep 1
spawn_with_rules xdotool type neofetch
sleep 1
spawn_with_rules xdotool key Return
spawn_with_rules herbstclient split right
RULES=( floating=off focus=off index='1' tag=1 )
spawn_with_rules alacrity
sleep 1
spawn_with_rules xdotool type la
sleep 1
spawn_with_rules xdotool key Return

# spawn an vifm in tag 2
RULES=( floating=off focus=off tag=2 )
spawn_with_rules alacritty -t Vifm
RULES=( tag=2 focus=on title=Vifm )
sleep 1
spawn_with_rules xdotool type vifm
sleep 1
spawn_with_rules xdotool key Return
# ~/.config/herbstluftwm/scripts/exec_on_tag.sh 4 ~/.local/bin/vifmrun # this script from official repo didn't work

# spawn an bottom (htop alternative) in tag 3
RULES=( floating=off focus=off tag=3 )
spawn_with_rules alacritty -t Bottom
sleep 1
RULES=( tag=3 focus=on title=Bottom )
spawn_with_rules xdotool type btm
sleep 1
spawn_with_rules xdotool key KP_Enter
```

**Comment by t-wissmann** [comment]
> ~/.config/herbstluftwm/scripts/exec_on_tag.sh 4 ~/.local/bin/vifmrun # this script from official repo didn't work

I don't know what the file `vifmrun` in your home directory is doing. But probably, it will not open a window. The following command works:
```
path/to/herbstluftwm/scripts/exec_on_tag.sh 4 alacritty -e vifm
```

As a general rule of thumb:
* **never** use `sleep` in these startup scripts, because otherwise your scripts will break almost always!
* **never** use `xdotool` in these startup scripts, because otherwise your scripts will break almost always!

Have you looked at the `RULES` section in the man page to understand what a window rule is?

Also note that `alacritty` contains `t` twice!

The following script probably does what you need:
```bash
#!/usr/bin/env bash

# Requirement: bash >= 4.0 (because of the usage of $BASHPID)

spawn_with_rules() {(
    # this rule also requires, that the client
    # sets the _NET_WM_PID property
    # herbstclient rule once pid=$BASHPID maxage=10 "${RULES[@]}"
    herbstclient rule once pid=$BASHPID "${RULES[@]}"
    exec "$@"
    ) &
}
herbstclient load 1 '(split horizontal:0.5:0 (clients max:1) (clients max:1))'

# spawn two terminals (alacritty and default terminal in two windows) in tag 1, left frame
RULES=( tag=1 index=0 )
spawn_with_rules alacritty --hold -e neofetch
RULES=( tag=1 index=1 )  # right-hand frame
spawn_with_rules alacritty --hold -e la

# spawn an vifm in tag 2
RULES=( tag=2 )
spawn_with_rules alacritty -e vifm

# spawn an bottom (htop alternative) in tag 3
RULES=( tag=3 )
spawn_with_rules alacritty -e btm
```

Thanks for mentioning bottom/btm, I didn't know it before and it looks quite nice! :-)

**Comment by RayZ0rr** [comment]
@t-wissmann Thanks for the answer. However, giving `neofetch` and `la` as command with alacritty will close it immediately after execution as there is no shell. Two workarounds: 
1) [alacritty/alacritty#1266](https://github.com/alacritty/alacritty/issues/1266)
2)
```bash
# Requirement: bash >= 4.0 (because of the usage of $BASHPID)

spawn_with_rules() {(
    # this rule also requires, that the client
    # sets the _NET_WM_PID property
    # herbstclient rule once pid=$BASHPID maxage=10 "${RULES[@]}"
    herbstclient rule once pid=$BASHPID "${RULES[@]}"
    exec "$@"
    ) &
}
herbstclient load 1 '(split horizontal:0.5:0 (clients max:1) (clients max:1))'

# spawn two terminals (alacritty and default terminal in two windows) in tag 1, left frame
RULES=( tag=1 index=0 )
spawn_with_rules alacritty
sleep 1
spawn_with_rules xdotool type neofetch
sleep 1
spawn_with_rules xdotool key Return
RULES=( tag=1 index=1 )  # right-hand frame
spawn_with_rules alacritty
sleep 1
spawn_with_rules xdotool type la
sleep 1
spawn_with_rules xdotool key Return

# spawn an vifm in tag 2
RULES=( tag=2 )
spawn_with_rules alacritty -e vifm

# spawn an bottom (htop alternative) in tag 3
RULES=( tag=3 )
spawn_with_rules alacritty
sleep 1
RULES=( tag=3 index=0 )
spawn_with_rules xdotool type btm
sleep 1
spawn_with_rules xdotool key Return
```

However, the last four lines: 
```bash
RULES=( tag=3 index=0 )
spawn_with_rules xdotool type btm
sleep 1
spawn_with_rules xdotool key Return
```
is executed in tag 1 alacrity and not tag 3. I know `-e` workaround exists but it would be great if I could send specific commands to specific tags with xdotool.

>  never use sleep in these startup scripts, because otherwise your scripts will break almost always!
    never use xdotool in these startup scripts, because otherwise your scripts will break almost always!

Sadly, when using `xdotool` the `sleep` command can become useful (necessary?) because otherwise multiple commands get executed at the same time (like `Return` can be pressed at `neo` of `neofetch`.

**Reply by t-wissmann** [reply]
If you want to keep 

> I know -e workaround exists but it would be great if I could send specific commands to specific tags with xdotool.

The flag `-e` is *not* a workaround. It does precisely what you are asking for: it executes the given command. If you don't want the terminal to close when the child process exits, then pass `--hold` to `alacritty`. Commands by the way are not "send to specific tags" (There is a difference between a window and a process. In your case alacritty provides the window, showing the output of the command passed to `-e`)

> > never use sleep in these startup scripts, because otherwise your scripts will break almost always!
> > never use xdotool in these startup scripts, because otherwise your scripts will break almost always!
> 
> Sadly, when using `xdotool` the `sleep` command can become useful (necessary?) because otherwise multiple commands get executed at the same time (like `Return` can be pressed at `neo` of `neofetch`.

No. I'm dead serious. You should **NEVER EVER** use them (for this purpose). They are neither useful nore necessary. If you have a multitasking system (like any system on the planet since the 90s), then it is fine if multiple commands are executed at the same time.
